src/libopensc/card-piv.c

```diff
@@ -467,82 +791,1819 @@ static struct sc_card_driver piv_drv = {
+		/* tell apdu.c to not do the chaining, let the SM get_apdu do it */
+		apdu.flags |= SC_APDU_FLAGS_SM_CHAINING;
+	}
+#endif
+	apdu.lc = sendbuflen;
+	apdu.datalen = sendbuflen;
+	apdu.data = sendbuf;
+
+	if (recvbuf && recvbuflen) {
+		apdu.le = (recvbuflen > 256) ? 256 : recvbuflen;
+		apdu.resplen = recvbuflen;
+	} else {
+		 apdu.le = 0;
+		 apdu.resplen = 0;
+	}
+	apdu.resp =  recvbuf;
+
+	/* with new adpu.c and chaining, this actually reads the whole object */
+	r = sc_transmit_apdu(card, &apdu);
+
+	/* TODO If r== SC_ERROR_SM_INVALID_CHECKSUM may need to restart SM */
+	/* adpu will not have sw1,sw2 set because sc_sm_single_transmit called sc_sm_stop, */
+	if (r < 0) {
+		sc_log(card->ctx, "Transmit failed");
+		goto err;
+	}
+
+	/* TODO Check this: with SM active, if other process interferes, 69 88 is returned */
+	if (apdu.sw1 == 0x69 && apdu.sw2 ==  0x88)
+		r = SC_ERROR_SM_INVALID_SESSION_KEY;
+	else
+		r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+
+	if (r < 0) {
+		sc_log(card->ctx,  "Card returned error ");
+		goto err;
+	}
+
+	r = apdu.resplen;
+
+err:
+	sc_unlock(card);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+
+#ifdef ENABLE_PIV_SM
+/* convert q as 04||x||y used in standard point formats to expanded leading
+ * zeros and concatenated X||Y as specified in SP80056A Appendix C.2
+ * Field-Element-to-Byte-String Conversion which
+ * OpenSSL has already converted X and Y to big endian and skipped leading
+ * zero bytes.
+ */
+static int Q2OS(int fsize, u8 *Q, size_t Qlen, u8 * OS, size_t *OSlen)
+{
+	size_t i;
+	size_t f = fsize/8;
+
+	i = (Qlen - 1)/2;
+
+	if (!OS || *OSlen < f * 2 || !Q || i > f)
+		return SC_ERROR_INTERNAL;
+
+	memset(OS, 0, f * 2);
+	/* Check this if x and y have leading zero bytes,
+	 * In UNCOMPRESSED FORMAT, x and Y must be same length, to tell when
+	 * one ends and the other starts */
+	memcpy(OS + f - i, Q + 1, i);
+	memcpy(OS + 2 * f - i, Q + f + 1, i);
+	*OSlen = f * 2;
+	return 0;
+}
+
+/*
+ * if needed, send VCI pairing code to card just after the
+ * SM key establishment. Called from piv_sm_open under same lock
+ */
+static int piv_send_vci_pairing_code(struct sc_card *card, u8 *paring_code)
+{
+	int r;
+	piv_private_data_t * priv = PIV_DATA(card);
+	sc_apdu_t plain;
+	sc_apdu_t sm_apdu;
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (priv->pin_policy & PIV_PP_VCI_WITHOUT_PC)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, SC_SUCCESS); /* Not needed */
+
+	if ((priv->pin_policy & PIV_PP_VCI_IMPL) == 0)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, SC_ERROR_NO_CARD_SUPPORT);
+
+	sc_format_apdu(card, &plain, SC_APDU_CASE_3_SHORT, 0x20, 0x00, 0x98);
+	plain.datalen = plain.lc = 8;
+	plain.data = paring_code;
+	plain.resp = NULL;
+	plain.resplen = plain.le = 0;
+
+	memset(&sm_apdu,0,sizeof(sm_apdu));
+	/* build sm_apdu and set alloc sm_apdu.resp */
+	r = piv_encode_apdu(card, &plain, &sm_apdu);
+	if (r < 0)
+		free(sm_apdu.resp);
+	LOG_TEST_RET(card->ctx, r, "piv_encode_apdu failed");
```

Comment:
I would rather see these checks rewritten to normal single condition `if (r < 0) { free, log, return }` to be clear what is done here. Or a least cover the `LOG_TEST_RET` under the `if` block.

Reply:
fixed